Firestore/core/src/firebase/firestore/remote/stream_operation.h

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_REMOTE_STREAM_OPERATION_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_REMOTE_STREAM_OPERATION_
+
+#include <chrono>  // NOLINT(build/c++11)
+#include <future>  // NOLINT(build/c++11)
+
+#include "Firestore/core/src/firebase/firestore/remote/grpc_operation.h"
+#include "Firestore/core/src/firebase/firestore/util/async_queue.h"
+#include "Firestore/core/src/firebase/firestore/util/status.h"
+#include "grpcpp/generic/generic_stub.h"
+#include "grpcpp/support/byte_buffer.h"
+
+namespace firebase {
+namespace firestore {
+namespace remote {
+
+class GrpcStream;
+
+/**
+ * An operation that notifies the corresponding `GrpcStream` on its completion.
+ *
+ * All created operations are always put on the GRPC completion queue (see
+ * `ExecuteOperation` factory function). Operation expects that once it's
+ * received back from the GRPC completion queue, `Complete()` will be called on
+ * it. `Complete` doesn't notify the observing stream immediately; instead, it
+ * schedules the notification on the Firestore async queue. If the stream
+ * doesn't want to be notified, it should call `UnsetObserver` on the operation.
+ *
+ * Operation is "self-owned"; operation deletes itself in its `Complete` method.
+ *
+ * Operation expects all GRPC objects pertaining to the current stream to remain
+ * valid until the operation comes back from the GRPC completion queue.
+ */
+class StreamOperation : public GrpcOperation {
```

Comment:
These and some other names are inconsistent. We have `GrpcOperation` with a sub-interface is `StreamOperation` where the various implementations are to be used only by the `GrpcStream` and not the higher-level portable `Stream`.

I think we should name things at the same level of abstraction consistently.

You might also want to rename `Stream` to `PersistentStream` to match the Javascript client and further separate it from the components at this level of abstraction.

Reply:
Question: I'm having doubts about the value of `GrpcOperation` as a pure interface. All the leaf classes inherit from `StreamOperation`, not directly from `GrpcOperation`. `StreamOperation` is now a public class (which I think is justified by its size). Perhaps it would be better to combine `GrpcOperation` and `StreamOperation` into one by removing the existing `GrpcOperation` (the pure interface) and renaming `StreamOperation` into `GrpcOperation` or `GrpcStreamOperation`?

Reply:
I had that in an earlier iteration. Did a quick restoration (see the last two commits) -- let me know what you think.